src/explorer/buildspecs.jl

```diff
@@ -0,0 +1,195 @@
+                "y" => Dict("field" => string(dffields[2]), "type" => "quantitative" )
+            ),
+            "width" => 400,
+            "height" => 400 
+        ),
+    )
+    return spec
+end
+
+function createspec_barplot(name, df, dffields)
+    datapart = getdatapart(df, dffields, :bar) #returns a list of dictionaries    
+    spec = Dict(
+        "name"  => name,
+        "VLspec" => Dict(
+            "\$schema" => "https://vega.github.io/schema/vega-lite/v2.0.json",
+            "description" => "plot for a specific component variable pair",
+            "title" => name,
+            "data"=> Dict("values" => datapart),
+            "mark" => "bar",
+            "encoding" => Dict(
+                "x" => Dict("field" => string(dffields[1]), "type" => "ordinal"),
+                "y" => Dict("field" => string(dffields[2]), "type" => "quantitative" )
+            ),
+            "width" => 400,
+            "height" => 400 
+        ),
+    )
+    return spec
+end
+
+function createspec_multilineplot(name, df, dffields)
+    datapart = getdatapart(df, dffields, :multiline) #returns a list of dictionaries    
+    spec = Dict(
+        "name"  => name,
+        "VLspec" => Dict(
+            "\$schema" => "https://vega.github.io/schema/vega-lite/v2.0.json",
+            "description" => "plot for a specific component variable pair",
+            "title" => name,
+            "data"=> Dict("values" => datapart),
+            "mark" => "line",
+            "encoding" => Dict(
+                "x" => Dict("field" => string(dffields[1]), "type" => "temporal", "timeUnit" => "year"),                
+                "y" => Dict("field" => string(dffields[3]), "type" => "quantitative" ),
+                "color" => Dict("field" => string(dffields[2]), "type" => "nominal", 
+                    "scale" => Dict("scheme" => "category20b"))
+            ),
+            "width" => 400,
+            "height" => 400 
+        ),
+    )
+    return spec
+end
+
+function createspec_singlevalue(name)
+
+    datapart = [];
+    spec = Dict(
+        "name" => name, 
+        "VLspec" => Dict()
+    )
+    return spec
+end
+
+function getdatapart(df, dffields, plottype::Symbol = :line)
+    datapart = [];
+
+    #loop over rows and create a dictionary for each row
+    if plottype == :multiline
+        for row in eachrow(df)
+            rowdata = Dict(string(dffields[1]) => Date(row[1]), string(dffields[3]) => row[3], 
+                string(dffields[2]) => row[2])
+            push!(datapart, rowdata)
+        end 
+    elseif plottype == :line
+        for row in eachrow(df)
+            rowdata = Dict(string(dffields[1])=> Date(row[1]), string(dffields[2]) => row[2])
+            push!(datapart, rowdata)
+        end 
+    else
+        for row in eachrow(df)
+            rowdata = Dict(string(dffields[1])=> row[1], string(dffields[2]) => row[2])
+            push!(datapart, rowdata)
+        end 
+    end
+    return datapart
+end
```

Comment:
This function is fine for now, but I suspect it is actually causing some of the performance issues we have. @lrennels can you open a new issue that we should revisit this eventually?

Reply:
All it does is create a closure with the relevant item from the spec list, and assign that function to the button. How would this create a performance problem?

The alternative approach is to assign the spec list to a global, but that prevents multiple windows from working properly. The closure approach embeds the state in the JS in the window.

Reply:
I think the ``eachrow`` approach is just fairly slow, and indexing into a row with e.g. ``row[1]`` is type unstable, so we are probably leaving a lot of performance on the table here. But we should just profile it and see whether this is where time is spent. If so, there are lots of ways we could make this specific function much faster.